django/template/defaulttags.py

```diff
@@ -1564,3 +1567,186 @@ def do_with(parser, token):
+            if context.template is None:
+                with context.bind_template(self):
+                    context.template_name = self.name
+                    return self._render(context)
+            else:
+                return self._render(context)
+
+
+class DefinePartialNode(Node):
+    def __init__(self, partial_name, inline, nodelist):
+        self.partial_name = partial_name
+        self.inline = inline
+        self.nodelist = nodelist
+
+    def render(self, context):
+        """Set content into context and return empty string"""
+        if self.inline:
+            return self.nodelist.render(context)
+        else:
+            return ""
+
+
+class RenderPartialNode(Node):
+    def __init__(self, partial_name, partial_mapping):
+        # Defer lookup of nodelist to runtime.
+        self.partial_name = partial_name
+        self.partial_mapping = partial_mapping
+
+    def render(self, context):
+        return self.partial_mapping[self.partial_name].render(context)
+
+
+@register.tag(name="partialdef")
+def partialdef_func(parser, token):
+    """
+    Declare a partial that can be used later in the template.
+
+    Usage::
+
+        {% partialdef partial_name %}
+        Partial content goes here
+        {% endpartialdef %}
+
+    Stores the nodelist in the context under the key "partial_contents" and can
+    be retrieved using the {% partial %} tag.
+
+    The optional ``inline`` argument will render the contents of the partial
+    where it is defined.
+    """
+    # Parse the tag
+    tokens = token.split_contents()
+
+    # check we have the expected number of tokens before trying to assign them
+    # via indexes
+    if len(tokens) not in (2, 3):
+        raise TemplateSyntaxError(
+            "%r tag requires 2-3 arguments" % token.contents.split()[0]
+        )
+
+    partial_name = tokens[1]
+
+    try:
+        inline = tokens[2]
+    except IndexError:
+        # the inline argument is optional, so fallback to not using it
+        inline = False
+
+    if inline and inline != "inline":
+        warnings.warn(
+            "The 'inline' argument does not have any parameters; "
+            "either use 'inline' or remove it completely.",
+            DeprecationWarning,
+        )
+
+    # Parse the content until the end tag
+    acceptable_endpartials = ("endpartialdef", f"endpartialdef {partial_name}")
+    nodelist = parser.parse(acceptable_endpartials)
+    endpartial = parser.next_token()
+    if endpartial.contents not in acceptable_endpartials:
+        parser.invalid_block_tag(endpartial, "endpartialdef", acceptable_endpartials)
+
+    # Store the partial nodelist in the parser.extra_data attribute,
+    parser.extra_data.setdefault("template-partials", {})
+    parser.extra_data["template-partials"][partial_name] = TemplateProxy(
+        nodelist, parser.origin, partial_name
+    )
+
+    return DefinePartialNode(partial_name, inline, nodelist)
+
+
+class SubDictionaryWrapper:
```

Comment:
Would this be better placed in the django.template.utils module? This could also use a test case covering the functionality?